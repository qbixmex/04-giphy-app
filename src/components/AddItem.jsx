import { useState } from "react";

const AddItem = ({ items, setItem }) => {

  const [ inputValue, setInputValue ] = useState('');

  const onInputChange = ({ target }) => {
    setInputValue(target.value);
  };

  const onAddItem = (event) => {
    event.preventDefault();
    if(inputValue) {
      setItem([ ...items, inputValue ]);
      setInputValue('');
    }
  };

  return (
    <form onSubmit={onAddItem}>
      <input
        className="form-control mt-4 mb-2"
        type="text"
        onChange={onInputChange}
        value={inputValue}
        placeholder='Search Gifs'
      />
      <div className="d-grid mb-3">
        <button
          type="submit"
          className="btn btn-primary"
          disabled={!inputValue && true}
        >
          Add
        </button>
      </div>
    </form>
  );
};

export default AddItem;
