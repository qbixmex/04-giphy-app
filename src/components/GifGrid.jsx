import { useState, useEffect } from "react";
import { getGifs } from "../helpers/getGifs";

const GifGrid = ({ category }) => {

  const [counter, setCounter] = useState(0);

  useEffect(() => {
    getGifs(category);
  }, []);

  return (
    <>
      <h2>{ category.name }</h2>
      <h3>{ counter }</h3>
      <button
        className="btn btn-primary"
        onClick={ () => setCounter(counter + 1) }
      >+1</button>
    </>
  );
};

export default GifGrid;
