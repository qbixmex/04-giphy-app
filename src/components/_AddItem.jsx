import { useState } from "react";
import PropTypes from 'prop-types';

const AddItem = ({ onNewItem }) => {

  const [inputValue, setInputValue] = useState('');

  const onInputChange = ({ target }) => {
    setInputValue(target.value);
  };

  const onSubmit = (event) => {
    event.preventDefault();
    const value = inputValue.trim();
    if (value.length <= 2) return;
    onNewItem(value);
    setInputValue('');
  };

  return (

    <form aria-label="form" onSubmit={onSubmit}>
      <div className="row mt-4 mb-4 g-2 justify-content-center align-items-center">
        <div className="col-12 col-md-6 col-lg-4 col-xl-3">
          <input
            className="form-control"
            type="text"
            onChange={onInputChange}
            value={inputValue}
            placeholder='Search Gifs'
          />
        </div>
        <div className="d-grid col-12 col-md-auto">
          <button
            type="submit"
            className="btn btn-primary d-inline"
            disabled={(inputValue.trim().length <= 2) && true}
          >
            Add
          </button>
        </div>
      </div>
    </form>
  );
};

AddItem.propTypes = {
  onNewItem: PropTypes.func.isRequired,
};

export default AddItem;
