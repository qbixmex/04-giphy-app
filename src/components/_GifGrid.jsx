import PropTypes from "prop-types";
import useFetchGifs from "../hooks/useFetchGifs";
import { GiphyItem, Spinner } from "./";

const GifGrid = ({ category }) => {

  const { gifs, isLoading } = useFetchGifs(category.name);

  return (
    <>
      <h2>{category.name}</h2>

      { isLoading && <Spinner color='blue' /> }

      <div className="row">
        {
          gifs.length > 0 && gifs.map(({ id, ...rest }) => (
            <GiphyItem key={id} {...rest} />
          ))
        }
      </div>
    </>
  );
};

GifGrid.propTypes = {
  category: PropTypes.object.isRequired,
};

export default GifGrid;
