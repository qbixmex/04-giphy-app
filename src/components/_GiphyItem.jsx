import PropTypes from 'prop-types';

const GiphyItem = ({ url, title }) => {
  return (
    <div className="col-xs-12 col-sm-6 col-lg-4">
      <div className="card mb-4">
        <img id="giphy-img" src={url} className="card-img-top" alt={title} />
        <div className="card-footer">
          <p id="footer" className="mb-0 fw-bold">{title}</p>
        </div>
      </div>
    </div>
  );
};

GiphyItem.propTypes = {
  url: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
};

export default GiphyItem;
