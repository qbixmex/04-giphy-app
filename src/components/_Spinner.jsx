import PropTypes from 'prop-types';

const colors = {
  blue: 'primary',
  gray: 'primary',
  cyan: 'info',
  green: 'success',
  yellow: 'warning',
  red: 'danger',
  light: 'light',
  dark: 'dark',
};

/**
 * Set a Spinner with bootstrap 5.1
 * @param {string} color Available Colors:
 * blue, gray, cyan, green, yellow, red, light, dark
 * @returns { JSX.Element }
 */
const Spinner = ({ color }) => {
  return (
    <div className={ `spinner-border text-${colors[color]}` } role="status">
      <span className="visually-hidden">Loading...</span>
    </div>
  );
};

Spinner.propTypes = {
  color: PropTypes.string,
};

Spinner.defaultProps = {
  color: 'dark',
};

export default Spinner;
