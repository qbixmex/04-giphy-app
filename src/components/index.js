import AddItem from "./_AddItem";
import GifGrid from "./_GifGrid";
import GiphyItem from "./_GiphyItem";
import Spinner from "./_Spinner";

export {
  AddItem,
  GifGrid,
  GiphyItem,
  Spinner,
};