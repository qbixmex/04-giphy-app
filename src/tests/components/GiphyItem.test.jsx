import { render } from "@testing-library/react";
import { GiphyItem } from "../../components";

describe('Test on <GiphyItem />', () => {

  const title = 'Dragon Ball Super';
  const url = 'https://giphy.com/dragon-ball-super.gif';

  test('Should match with snapshot', () => {
    const { container } = render(
      <GiphyItem title={ title } url={ url } />
    );
    expect(container).toMatchSnapshot();
  });

  test('Should has an image with alt and url', () => {
    const { container } = render(
      <GiphyItem title={ title } url={ url } />
    );

    const { src, alt } = container.querySelector('#giphy-img');

    expect(src).toBe(url);
    expect(alt).toBe(title);

  });

  test('Should has a footer title', () => {
    const { container } = render(
      <GiphyItem title={ title } url={ url } />
    );

    const pTitle = container.querySelector('#footer');

    expect(pTitle.innerHTML).toBe(title);
  });

});