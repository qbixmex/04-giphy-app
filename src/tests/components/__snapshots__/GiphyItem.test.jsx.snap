// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <GiphyItem /> Should match with snapshot 1`] = `
<div>
  <div
    class="col-xs-12 col-sm-6 col-lg-4"
  >
    <div
      class="card mb-4"
    >
      <img
        alt="Dragon Ball Super"
        class="card-img-top"
        id="giphy-img"
        src="https://giphy.com/dragon-ball-super.gif"
      />
      <div
        class="card-footer"
      >
        <p
          class="mb-0 fw-bold"
          id="footer"
        >
          Dragon Ball Super
        </p>
      </div>
    </div>
  </div>
</div>
`;
